### PR TITLE
Split run file

### LINKED ISSUE #144: Helper function to split run json into smaller files
*This is for internal use only; if you'd like to open an issue or request a new feature, please open a bug or enhancement issue*

**Section 1: Design details**

**Relevant background**

Closes #82. Create a new helper function in `json_utils.py` that will split the run json into smaller files for imaging.

**Design overview**

If provided a run_json with 10 fovs and list [4,3,3], will produce 3 new jsons.
* run_json_part1 = [fov1, fov2, fov3, fov4]
* run_json_part2 = [fov5, fov6, fov7,]
* run_json_part3 = [fov8, fov9, fov10]

The function will raise an error if the list does not sum to the number of fovs given in a run file.

```
def split_run_file(tma_dir, run_file_name, file_split: List):
    json_path = os.path.join(tma_dir, run_file_name)
    full_json = read_json_file(json_path, encoding='utf-8')

    # check list is valid FOV split
    if not sum(file_split)==len(full_json['fovs']):
        raise ValueError("Sum of the provided list does not match the number of FOVs in the run file.")
    
    # split the run json into smaller files
    start=0
    for i in len(file_split):
        json_i = copy.deepcopy(full_json)
        stop = start+file_split[i]
        json_i['fovs'] = json_i['fovs'][start:stop]
        start = start+file_split[i]

        save_path = os.path.join(tma_dir, run_file_name.split('.json')[0] + '_part' + str(i+1) + '.json)
        json_utils.write_json_file(save_path, json_i, encoding='utf-8')

```

Will be implemented in [2_create_tmab_mibi_run notebook](https://github.com/angelolab/toffy/blob/ac80af8ddca601ce91065de60bb2be4d70171262/templates/2_create_tma_mibi_run.ipynb).

**Required inputs**

* path to tma directory
* name of the run json file
* list detailing how many fovs will be in each new json file, i.e. [12, 12, 10]

**Output files**

Specified number of new json files. 

**Section 2: Implementation details**

Once you have completed section 1, please tag the relevant parties and iterate on the initial design details until everyone is satisfied. Then, 
proceed to section 2

**Control flow**

Provide additional, more granular details (if necessary) about how the proposed coding logic will be put together

**Milestones and timeline**

List each of the major components of the project, and provide an estimated completion date for each one.

### PR DESCRIPTION
**Pull request by camisowers**
**If you haven't already, please read through our [contributing guidelines](https://ark-analysis.readthedocs.io/en/latest/_rtd/contributing.html) before opening your PR**

**What is the purpose of this PR?**

Closes #144. Added a helper function to json_utils.py that splits the run json into smaller files for imaging. Also updated the notebooks to use the new function.

### CODE REVIEW

File: toffy/json_utils.py [hunk 1]
```diff
@@ -114,3 +114,35 @@ def write_json_file(json_path, json_object, encoding=None):
...
+    Returns:
+        saves the new json files to the base_dir """
+
+    json_path = os.path.join(run_dir, run_file_name)
+    full_json = read_json_file(json_path, encoding='utf-8')
+
+    # check list is valid FOV split
+    if not sum(file_split) == len(full_json['fovs']):
+        raise ValueError(
+            "Sum of the provided list does not match the number of FOVs in the run file.")
```
Comment (ngreenwald) [inline toffy/json_utils.py:136]: Add a note saying what the sum and lengths are

File: toffy/json_utils_test.py [hunk 2]
```diff
@@ -130,3 +130,32 @@ def test_write_json_file():
...
+        # test successful json split
+        good_split = [2, 4, 2]
+        json_utils.split_run_file(tmp_dir, run_file_name, good_split)
+
+        # check the new files exist
+        new_data = {}
+        for i in list(range(1, len(good_split)+1)):
+            new_json = os.path.join(tmp_dir, 'test_part' + str(i) + '.json')
+            new_data['test_part{0}'.format(i)] = json_utils.read_json_file(new_json, 'utf-8')
+            assert os.path.exists(new_json)
```
Comment (ngreenwald) [inline toffy/json_utils_test.py:156]: This assert will never fail, if the file wasn't created then the previous line reading in the file will fail. I think it's fine to remove it

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good, just a few tweaks

File: templates/2_create_tiled_mibi_run.ipynb [hunk 3]
```diff
@@ -18,7 +18,7 @@
   },
```
Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:18]: <p>Why is this cell needed?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/145/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (ngreenwald) [inline templates/2_create_tiled_mibi_run.ipynb:18]: <p>This should default to reading in the tiled json that was just created above, not the original run file</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/145/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

File: templates/2_create_tma_mibi_run.ipynb [hunk 4]
```diff
@@ -294,6 +294,44 @@
     "# save remapped_fov_regions\n",
```
Comment (ngreenwald) [inline templates/2_create_tma_mibi_run.ipynb:294]: <p>Same comments as previous notebook</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/angelolab/toffy/pull/145/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Great, and did you run the notebooks to make sure there's no syntax errors?

**Comment by camisowers** [comment]
> Great, and did you run the notebooks to make sure there's no syntax errors?

Yes all good!

**Review by ngreenwald — APPROVED** [review]

**Pull request merged**